File: src/shared/utils/security.ts
/**
 * @file security.ts
 * @description A utility module providing security-related functions and constants for the Femtosense Voice Command Generation system,
 * ensuring secure API communication, data handling, and authentication.
 *
 * Requirements addressed:
 * - API Authentication (Technical Specification/6.1 AUTHENTICATION AND AUTHORIZATION)
 * - Data Security (Technical Specification/6.2 DATA SECURITY)
 * - Security Protocols (Technical Specification/6.3 SECURITY PROTOCOLS)
 */

import crypto from 'crypto';
import { ApiResponse, createSuccessResponse, createErrorResponse, HttpStatusCode } from '../interfaces/api-response.interface';
import { API_ENDPOINTS } from '../constants/api-endpoints';

// Constants
export const ENCRYPTION_KEY = process.env.ENCRYPTION_KEY || 'default-key-replace-in-production';
export const API_KEY_HEADER = 'x-api-key';

/**
 * Encrypts the API key using AES-256-GCM encryption for secure storage and transmission.
 * @param apiKey The API key to encrypt
 * @returns The encrypted API key as a string
 */
export function encryptApiKey(apiKey: string): string {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', Buffer.from(ENCRYPTION_KEY, 'hex'), iv);
  const encrypted = Buffer.concat([cipher.update(apiKey, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  return Buffer.concat([iv, encrypted, tag]).toString('hex');
}

/**
 * Decrypts an encrypted API key for use in API calls.
 * @param encryptedKey The encrypted API key
 * @returns The decrypted API key as a string
 */
export function decryptApiKey(encryptedKey: string): string {
  const buffer = Buffer.from(encryptedKey, 'hex');
  const iv = buffer.slice(0, 12);
  const tag = buffer.slice(-16);
  const encryptedText = buffer.slice(12, -16);
  const decipher = crypto.createDecipheriv('aes-256-gcm', Buffer.from(ENCRYPTION_KEY, 'hex'), iv);
  decipher.setAuthTag(tag);
  const decrypted = decipher.update(encryptedText) + decipher.final('utf8');
  return decrypted;
}

/**
 * Generates a unique request ID for API call tracking and debugging.
 * @returns A unique request ID as a string
 */
export function generateRequestId(): string {
  return crypto.randomUUID();
}

/**
 * Validates an API key against the service endpoints.
 * @param apiKey The API key to validate
 * @returns A Promise resolving to an ApiResponse indicating if the API key is valid
 */
export async function validateApiKey(apiKey: string): Promise<ApiResponse<boolean>> {
  const requestId = generateRequestId();
  try {
    // This is a mock implementation. In a real-world scenario, you would make an API call to validate the key.
    const isValid = apiKey.length === 32; // Simple check for demonstration purposes
    if (isValid) {
      return createSuccessResponse(true, requestId);
    } else {
      return createErrorResponse({
        code: HttpStatusCode.UNAUTHORIZED,
        message: 'Invalid API key',
      }, requestId);
    }
  } catch (error) {
    return createErrorResponse({
      code: HttpStatusCode.INTERNAL_SERVER_ERROR,
      message: 'Error validating API key',
      details: { error: error.message },
    }, requestId);
  }
}

/**
 * Sanitizes user input to prevent injection attacks.
 * @param input The user input to sanitize
 * @returns The sanitized input string
 */
export function sanitizeInput(input: string): string {
  // Remove any HTML tags
  let sanitized = input.replace(/<[^>]*>?/gm, '');
  // Escape special characters
  sanitized = sanitized.replace(/[&<>"']/g, (char) => {
    const escapeChars: { [key: string]: string } = {
      '&': '&amp;',
      '<': '&lt;',
      '>': '&gt;',
      '"': '&quot;',
      "'": '&#39;'
    };
    return escapeChars[char] || char;
  });
  return sanitized;
}

/**
 * A class that provides secure API communication methods.
 */
export class SecureAPIClient {
  private apiKey: string;

  constructor(apiKey: string) {
    this.apiKey = encryptApiKey(apiKey);
  }

  /**
   * Makes a secure API request with proper authentication and error handling.
   * @param endpoint The API endpoint to call
   * @param method The HTTP method to use
   * @param data Optional data to send with the request
   * @returns A Promise resolving to a typed ApiResponse
   */
  async request<T>(endpoint: string, method: string, data?: unknown): Promise<ApiResponse<T>> {
    const requestId = generateRequestId();
    try {
      const decryptedKey = decryptApiKey(this.apiKey);
      const headers = {
        [API_KEY_HEADER]: decryptedKey,
        'Content-Type': 'application/json',
      };

      const response = await fetch(endpoint, {
        method,
        headers,
        body: data ? JSON.stringify(data) : undefined,
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const result = await response.json();
      return createSuccessResponse<T>(result, requestId);
    } catch (error) {
      return createErrorResponse({
        code: HttpStatusCode.INTERNAL_SERVER_ERROR,
        message: 'Error making API request',
        details: { error: error.message },
      }, requestId);
    }
  }
}

// Example usage:
// const apiClient = new SecureAPIClient('your-api-key-here');
// const response = await apiClient.request<YourResponseType>(API_ENDPOINTS.gpt.variations, 'POST', { prompt: 'Your prompt here' });

File: src/backend/src/utils/security.py
import os
import hashlib
import secrets
from typing import Optional
from cryptography.fernet import Fernet
from ..config.app_config import AppConfig
from .logger import logger, log_decorator

# Constants for API key header and encryption key environment variable
API_KEY_HEADER: str = 'x-api-key'
ENCRYPTION_KEY_ENV_VAR: str = 'FEMTOSENSE_ENCRYPTION_KEY'

class SecurityException(Exception):
    """Custom exception for security-related errors."""
    pass

@log_decorator("INFO")
def get_encryption_key() -> bytes:
    """
    Retrieves the encryption key from environment variables, ensuring it exists and is valid.

    Returns:
        bytes: The encryption key

    Raises:
        SecurityException: If the encryption key is not set or invalid

    Requirements addressed:
    - Data Security (Technical Specification/6.2 DATA SECURITY)
    - Security Protocols (Technical Specification/6.3 SECURITY PROTOCOLS)
    """
    encryption_key = os.environ.get(ENCRYPTION_KEY_ENV_VAR)
    if not encryption_key:
        logger.error(f"Encryption key not found in environment variable: {ENCRYPTION_KEY_ENV_VAR}")
        raise SecurityException("Encryption key not set in environment variables")
    
    try:
        return encryption_key.encode()
    except Exception as e:
        logger.error(f"Invalid encryption key: {str(e)}")
        raise SecurityException("Invalid encryption key")

@log_decorator("INFO")
def encrypt_sensitive_data(data: str) -> str:
    """
    Encrypts sensitive data using Fernet symmetric encryption.

    Args:
        data (str): The sensitive data to encrypt

    Returns:
        str: Encrypted data string

    Raises:
        SecurityException: If encryption fails

    Requirements addressed:
    - Data Security (Technical Specification/6.2 DATA SECURITY)
    """
    try:
        key = get_encryption_key()
        cipher_suite = Fernet(key)
        encrypted_data = cipher_suite.encrypt(data.encode())
        return encrypted_data.decode()
    except Exception as e:
        logger.error(f"Encryption failed: {str(e)}")
        raise SecurityException("Failed to encrypt sensitive data")

@log_decorator("INFO")
def decrypt_sensitive_data(encrypted_data: str) -> str:
    """
    Decrypts previously encrypted sensitive data.

    Args:
        encrypted_data (str): The encrypted data string

    Returns:
        str: Decrypted data string

    Raises:
        SecurityException: If decryption fails

    Requirements addressed:
    - Data Security (Technical Specification/6.2 DATA SECURITY)
    """
    try:
        key = get_encryption_key()
        cipher_suite = Fernet(key)
        decrypted_data = cipher_suite.decrypt(encrypted_data.encode())
        return decrypted_data.decode()
    except Exception as e:
        logger.error(f"Decryption failed: {str(e)}")
        raise SecurityException("Failed to decrypt sensitive data")

@log_decorator("INFO")
def validate_api_key(api_key: str) -> bool:
    """
    Validates an API key against stored secure hash.

    Args:
        api_key (str): The API key to validate

    Returns:
        bool: True if valid, False otherwise

    Requirements addressed:
    - API Authentication (Technical Specification/6.1 AUTHENTICATION AND AUTHORIZATION)
    """
    stored_hash = AppConfig.get_api_key_hash()
    if not stored_hash:
        logger.error("No API key hash found in configuration")
        return False
    
    hashed_key = hashlib.sha256(api_key.encode()).hexdigest()
    return secrets.compare_digest(hashed_key, stored_hash)

@log_decorator("DEBUG")
def generate_secure_filename(base_name: str) -> str:
    """
    Generates a cryptographically secure filename for storing sensitive files.

    Args:
        base_name (str): The base name for the file

    Returns:
        str: Secure filename

    Requirements addressed:
    - Data Security (Technical Specification/6.2 DATA SECURITY)
    - Security Protocols (Technical Specification/6.3 SECURITY PROTOCOLS)
    """
    random_suffix = secrets.token_hex(8)
    return f"{base_name}_{random_suffix}"

@log_decorator("DEBUG")
def sanitize_input(input_str: str) -> str:
    """
    Sanitizes user input to prevent injection attacks.

    Args:
        input_str (str): The input string to sanitize

    Returns:
        str: Sanitized input string

    Requirements addressed:
    - Security Protocols (Technical Specification/6.3 SECURITY PROTOCOLS)
    """
    # Remove potentially dangerous characters
    sanitized = ''.join(char for char in input_str if char.isalnum() or char in [' ', '-', '_', '.'])
    # Escape special characters
    sanitized = sanitized.replace("'", "''")
    return sanitized

class SecureAPIClient:
    """
    A class that provides secure API communication methods.

    Requirements addressed:
    - API Authentication (Technical Specification/6.1 AUTHENTICATION AND AUTHORIZATION)
    - Data Security (Technical Specification/6.2 DATA SECURITY)
    """

    def __init__(self, api_key: str):
        """
        Initializes the SecureAPIClient with an API key

        Args:
            api_key (str): The API key for authentication

        Raises:
            SecurityException: If the API key is invalid
        """
        if not validate_api_key(api_key):
            logger.error("Invalid API key provided")
            raise SecurityException("Invalid API key")
        self._encrypted_api_key = encrypt_sensitive_data(api_key)

    @log_decorator("INFO")
    def request(self, endpoint: str, method: str, data: Optional[dict] = None) -> dict:
        """
        Makes a secure API request with proper authentication and error handling.

        Args:
            endpoint (str): The API endpoint
            method (str): The HTTP method (GET, POST, etc.)
            data (Optional[Dict]): The request payload

        Returns:
            Dict: API response data

        Raises:
            SecurityException: If the request fails or returns an error

        Requirements addressed:
        - API Authentication (Technical Specification/6.1 AUTHENTICATION AND AUTHORIZATION)
        - Security Protocols (Technical Specification/6.3 SECURITY PROTOCOLS)
        """
        import requests  # Import here to avoid circular dependencies

        try:
            api_key = decrypt_sensitive_data(self._encrypted_api_key)
            headers = {API_KEY_HEADER: api_key}
            
            response = requests.request(method, endpoint, headers=headers, json=data)
            response.raise_for_status()
            
            return response.json()
        except requests.RequestException as e:
            logger.error(f"API request failed: {str(e)}")
            raise SecurityException(f"API request failed: {str(e)}")

# Initialize the logger for this module
logger = logger.getChild(__name__)